### ISSUE TITLE
BSOD on Windows using the COM port driver in xScheduler

### ISSUE THREAD

**Issue by PKingRoch** [smeighan/xLights#1292]
I'm able to consistently crash my system using a COM port for either LOR or DMX networks.  After one or two songs playing, I receive the following BSOD error.

The computer has rebooted from a bugcheck.  The bugcheck was: 0x00000139 (0x0000000000000003, 0xffffcd04a39684c0, 0xffffcd04a3968418, 0x0000000000000000). A dump was saved in: C:\WINDOWS\MEMORY.DMP. Report Id: 5413852e-18e0-47f1-9f50-3d62c85cdac4.

The dump is no longer on my system, but I can recreate this issue for you if you need it.  I was using the LOR USB 485 high speed adapter as well as the Holiday Coro ActiDongle #53/Vs - Enttec Pro Compatible DMX dongle.  Both produced the same BDOD.

Windows 10 v1803

**Comment by keithsw1111** [comment]
Which xLights version. Change was made to reduce the chances of this happening in a recent release

**Comment by AzGilrock** [comment]
Yeah if this is still happening on the latest version then I give up and I'm gonna say we can't fix it.

**Comment by PKingRoch** [comment]
This is happening on 2018.37.  If I can get a dump, it might help.  Let me investigate.

**Comment by keithsw1111** [comment]
Bump the serial speed up as high as you can.

I assume you are using the pixies.

Other options include lowering the frame rate, reducing the number of pixels on the controllers.

Like Gil I am beginning to see why we avoided the pixies in the past.

**Comment by AzGilrock** [comment]
Problem is a blue screen crash occurs at the driver level.  I should not be able to crash the driver no matter what I do in xLights code.  I spent way more hours than I ever wanted trying to fix this and other issues.  I finally have it where my blue screen crash was not occurring anymore.  I not spending anymore time on it.  Forget LOR...sell all your LOR adapters....or run everything from LOR.

**Comment by PKingRoch** [comment]
That's what I'm trying to point out, this has nothing to do with LOR.  I could reproduce this trying to drive DMX across a COM port as well.  Where does the COM port code come from that is being used?  Is there a later pull that can be used?  Honestly, I'm just trying to help here.  

[This URL](https://docs.microsoft.com/en-us/windows-hardware/drivers/debugger/bug-check---bug-check-0x139-kernel-security-check-failure) helps with figuring out what went wrong.

A LIST_ENTRY was corrupted (for example, a double remove). For more information, see the following Cause section.

CAUSE
LIST_ENTRY corruption can be difficult to track down and this bug check, indicates that an inconsistency has been introduced into a doubly-linked list (detected when an individual list entry element is added to or removed from the list). Unfortunately, the inconsistency is not necessarily detected at the time when the corruption occurred, so some detective work may be necessary to identify the root cause.
Common causes of list entry corruption include:
A driver has corrupted a kernel synchronization object, such as a KEVENT (for example double initializing a KEVENT while a thread was still waiting on that same KEVENT, or allowing a stack-based KEVENT to go out of scope while another thread was using that KEVENT). This type of bug check typically occurs in nt!Ke* or nt!Ki* code. It can happen when a thread finishes waiting on a synchronization object or when code attempts to put a synchronization object in the signaled state. Usually, the synchronization object being signaled is the one that has been corrupted. Sometimes, Driver Verifier with special pool can help track down the culprit (if the corrupted synchronization object is in a pool block that has already been freed).
A driver has corrupted a periodic KTIMER. This type of bug check typically occurs in nt!Ke* or nt!Ki* code and involves signaling a timer, or inserting or removing a timer from a timer table. The timer being manipulated may be the corrupted one, but it might be necessary to inspect the timer table with !timer (or manually walking the timer list links) to identify which timer has been corrupted. Sometimes, Driver Verifier with special pool can help track down the culprit (if the corrupted KTIMER is in a pool block that has already been freed).
A driver has mismanaged an internal LIST_ENTRY-style linked list. A typical example would be calling RemoveEntryList twice on the same list entry without reinserting the list entry between the two RemoveEntryList calls. Other variations are possible, such as double inserting an entry into the same list.
A driver has freed a data structure that contains a LIST_ENTRY without removing the data structure from its corresponding list, causing corruption to be detected later when the list is examined after the old pool block has been reused.
A driver has used a LIST_ENTRY-style list in a concurrent fashion without proper synchronization, resulting in a torn update to the list.
In most cases, you can identify the corrupted data structure by walking the linked list both forward and backwards (the dl and dlb commands are useful for this purpose) and comparing the results. Where the list is inconsistent between a forward and backward walk is typically the location of the corruption. Since a linked list update operation can modify the list links of a neighboring element, you should look at the neighbors of a corrupted list entry closely, as they may be the underlying culprit.
Because many system components internally utilize LIST_ENTRY lists, various types of resource mismanagement by a driver using system APIs might cause linked list corruption in a system-managed linked list.

**Comment by keithsw1111** [comment]
Here’s the thing we know why it occurs. 

It occurs when we close the comm port before it has cleared all the data in the send buffer. 

Problem is we have done everything in our power to manage this. 

We tried purging it
We tried limiting the data we send if there is already data there
We tried giving it time to clear. 

All have failed. 

The bug you are seeing is not our code. The driver is crashing and it shouldn’t be. Only Microsoft can fix it from here. 

There are things you can do. 

Increase the speed of the port. That way it clears quicker. 
Reduce pixels. That way we send less data. 
Avoid effects on these models in the last part of the song. 

Gil has spent a lot of time getting as far as this and we are pretty sure this is the facts of the situation.

**Comment by AzGilrock** [comment]
Look at your "cause" list.  It is full of "A driver has....".   That "driver" they are talking about is NOT our code.

**Comment by PKingRoch** [comment]
Let me try and help here.  Where in your code are you working with the COM port?  I could build this debug and take a look at it.  I assume you have a document for setting up a dev environment, can you point me to it?  If it is the Microsoft driver, then I'll see about getting it fixed.  I work for Microsoft.

**Comment by PKingRoch** [comment]
Holiday Coro ActiDongle #53/V2 - Enttec Pro Compatible DMX dongle.  It a USB serial port and uses FTDI according to my device manager.  I did get an updated driver when I investigated the device just now.  Let me do some additional testing later and see if it still fails.

**Comment by AzGilrock** [comment]
You made it sound like its crashing while you are still trying to send data.  I've only had the BSOD's happen when trying to close the port.  We've experimented with using the TXCLEAR, RXCLEAR, TXPURGE, and RXPURGE and that didn't help much.  I finally tried adding code that polls TxEmpty and doesn't close the port until it says its empty.  I found that if you pumped too much data to a Pixie we were getting a ton of extra data stacking up and I saw it take 25 seconds to clear the TX buffer.  So I added code for the LOROutput that prevents it from building a stackup of extra frames.  Now we drop frames if the COM port can't keep up.  I still didn't think a driver should crash all because it still had data in the transmit buffer when we told it to close.

**Comment by AzGilrock** [comment]
By the way most of that serial code was written way before Keith and I joined xLights.  Probably Matt Brown code.  Keith refactored the outputs so although you may see his name on a lot of it that's only cause he moved and reorganized it.  Then I've really only touched the LOROptimised code and the LOR controller classes I had to create to do that work.

**Comment by PKingRoch** [comment]
Well, interesting tests today.  As I mentioned, the FTDI driver updated on my system this morning.  I'm now using version 2.12.28.0 dated 8/16/2017.  Right now, the LOR dongle seems to be working with no crash. The DMX dongle however now no longer outputs to the lights, but I couldn't crash it either in communicating with it.

I was able to clone the code and get everything to build 64-bit debug.  I see what you mean about there really isn't much in your code.  I saw the 200ms loop and your flush; nothing in your code for this.  While there may be something in the serial_win32.cpp file around using Overlapped IO causing this, without a repro, it's pretty hard to see what the heck is going wrong.  

I could roll back the driver, but why.  At this point, I don't need a serial port.  I've moved everything to E1.31.  I really just wanted to give you a heads up that I found this as I was doing some of my testing this year.

Feel free to close it.  Thanks for the indulgence.

**Comment by keithsw1111** [comment]
Can you add a link to the updated driver you used.

**Comment by PKingRoch** [comment]
I don't have one.  I updated through "Update Driver" on the FTDI "driver" tab of the "USB Serial Port Driver" properties dialog you get when you go through Device Manager and right click on the serial port.  It didn't appear to be a driver that was pushed, so anyone with this problem will need to get the driver the same way.  I also took a quick look on the FTDI website and I didn't see, at least not from a cursory look.
